direct/nn/cirim/cirim_engine.py

```diff
@@ -0,0 +1,518 @@
+        regularizer_fns: Optional[Dict[str, Callable]] = None,
+    ) -> DoIterationOutput:
+
+        # loss_fns can be done, e.g. during validation
+        if loss_fns is None:
+            loss_fns = {}
+
+        if regularizer_fns is None:
+            regularizer_fns = {}
+
+        # The first input_image in the iteration is the input_image with the mask applied and no first hidden state.
+        loss_dicts = []
+        regularizer_dicts = []
+
+        data = dict_to_device(data, self.device)
+
+        # sensitivity_map of shape (batch, coil, [slice], height,  width, complex=2)
+        sensitivity_map = data["sensitivity_map"]
+
+        if "noise_model" in self.models:
+            raise NotImplementedError()
+
+        # Some things can be done with the sensitivity map here, e.g. apply a u-net
+        if "sensitivity_model" in self.models:
+            # Move channels to first axis
+            sensitivity_map = data["sensitivity_map"].permute(
+                (0, 1, 4, 2, 3) if self.ndim == 2 else (0, 1, 5, 2, 3, 4)
+            )  # shape (batch, coil, complex=2, [slice], height,  width)
+
+            sensitivity_map = self.compute_model_per_coil("sensitivity_model", sensitivity_map).permute(
+                (0, 1, 3, 4, 2) if self.ndim == 2 else (0, 1, 3, 4, 5, 2)
+            )  # has channel last: shape (batch, coil, [slice], height,  width, complex=2)
+
+        # The sensitivity map needs to be normalized such that
+        # So \sum_{i \in \text{coils}} S_i S_i^* = 1
+
+        sensitivity_map_norm = torch.sqrt(
+            ((sensitivity_map ** 2).sum(self._complex_dim)).sum(self._coil_dim)
+        )  # shape (batch, [slice], height, width)
+        sensitivity_map_norm = sensitivity_map_norm.unsqueeze(self._coil_dim).unsqueeze(self._complex_dim)
+        data["sensitivity_map"] = T.safe_divide(sensitivity_map, sensitivity_map_norm)
+
+        with autocast(enabled=self.mixed_precision):
+            output_image = self.model(
+                masked_kspace=data["masked_kspace"],
+                sampling_mask=data["sampling_mask"],
+                sensitivity_map=data["sensitivity_map"],
+            )
+
+            loss_dict = {k: torch.tensor([0.0], dtype=data["target"].dtype).to(self.device) for k in loss_fns.keys()}
+            regularizer_dict = {
+                k: torch.tensor([0.0], dtype=data["target"].dtype).to(self.device) for k in regularizer_fns.keys()
+            }
+
+            for key, value in loss_dict.items():
+                loss_ = sum(self.process_loss(loss_fns[key], reduction="mean", output_image=output_image, **data))
+                loss_dict[key] = value + loss_
+
+            for key, value in regularizer_dict.items():
+                regularizer_dict[key] = value + sum(
+                    self.process_loss(regularizer_fns[key], output_image=output_image, **data)
+                )
+
+            loss = sum(loss_dict.values()) + sum(regularizer_dict.values())
```

Comment:
I suggest removing the process_loss method and do something similar with `rim_engine` (which also has a multi-time output). Check https://github.com/NKI-AI/direct/blob/d874e8d983dd854f086c45844136cfdef3925342/direct/nn/rim/rim_engine.py#L130-L157. You can do the same as there but just weight the `output_image_iter`.
I am worried that the nan may be caused by that function.